enhancements/ingress/dynamic-config-manager.md

```diff
@@ -0,0 +1,469 @@
+servers dynamically, add backends dynamically, and update certificates
+dynamically.
+
+### Risks and Mitigations
+
+Dynamic Config Manager has some known defects:
+
+- [OCPBUGS-7466 No prometheus haproxy metrics present for route created by dynamicConfigManager](https://issues.redhat.com/browse/OCPBUGS-7466)
+- [OCPBUGSM-20868 Sticky cookies take long to start working if config manager is enabled](https://issues.redhat.com/browse/OCPBUGSM-20868)
+- [NE-1815 Fix implementation gaps discovered during the smoke tests](https://issues.redhat.com/browse/NE-1815)
+
+We intend to fix all of these defects before enabling DCM by default as GA.
+However, DCM could have additional, unknown defects.  We will mitigate this risk
+through E2E tests, payload tests, and working with partners to test DCM.
+
+### Drawbacks
+
+#### Memory overhead
+
+DCM also has some up-front memory overhead.  For this reason, DCM will be
+initially enabled with a minimum configuration of 1 pre-allocated server slot
+per backend and without blueprint routes.  Eventually, DCM will be enhanced to
+use newer HAProxy capabilities to perform more configuration without
+pre-allocating servers or backends, thus preventing reloads in more cases
+without adding any additional up-front memory overhead.
+
+#### Redundancy with respect to Istio/Envoy
+
+DCM is similar in concept to [Envoy
+xDS](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/operations/dynamic_configuration).
+Both enable a control plane (openshift-router or Istio) to configure a data
+plane (HAProxy or Envoy, respectively) without restarting processes.  Investing
+in DCM could be considered a redundant effort when Istio/Envoy already exists
+and avoids the problem.  However, many customers depend on OpenShift router for
+its performance and reliability and will continue using it indefinitely.  For
+these customers, DCM could improve OpenShift router's performance and
+reliability without forcing a change in APIs (from Route API to Gateway API) or
+proxies (from HAProxy to Envoy).
+
+## Design Details
+
+### Open Questions [optional]
+
+#### Can we use HAProxy's control socket to add backends and certificates dynamically?
+
+Answer: **?**.
+
+We need to determine what kind of configuration changes cannot be implemented in
+DCM and either document these limitations or work with HAProxy upstream to
+enhance HAProxy's management capabilities and remove these limitations.
+
+#### Can we configure Dynamic Config Manager with 0 pre-allocated servers?
+
+Answer: **Probably no, but need verification**.
+
+One option could be to enable DCM initially with 0 pre-allocated server slots
+per backend.  If this worked, it would be useful to avoid reloads for scale-in,
+or for scale-out following scale-in.  However, we found that DCM still requires
+a reload for scale-in of a server that doesn't use a pre-allocated server slot.
+(TODO: This needs to be double-checked.)
```

Comment:
I did see stopped not pre-allocated servers. So, scale ins should still be dynamic.

Reply:
Btw even the route deletion doesn't lead to a reload if DCM is enabled. All of the servers of the deleted backend will be disabled but the backend will stay in `show backend` output.

Reply:
This raises the question of whether we need to log additional operational debug information somewhere. With DCM enabled, the haproxy.config becomes stale, so relying solely on its contents during debugging could lead to misleading conclusions.

Reply:
Right. I had [a user story for this](https://issues.redhat.com/browse/NE-1823) but didn't have enough time to implement something for this iteration.

Reply:
> I tried to scale a deployment from 4 replicas (`3 pre-allocated slots UP + 1 dynamic slot UP`) to 1. Here is the output of `show servers state`:

I'm confused.  I've been saying "pre-allocated" to describe the slots that we pre-allocate in `haproxy.config` for DCM to use to add new server endpoints to a backend.  What do you mean when you say "3 pre-allocated slots"?  Are these 3 pre-allocated slots for DCM, or are these 3 `server` lines for endpoints that are defined in `haproxy.config`?  

Are you saying that I can create a route with 3 endpoints (which get written to `haproxy.config`), and then scale in to 1 endpoint, and no reload is required?  

If I have a route with 1 pod endpoint (written to `haproxy.config`), delete the pod, then recreate the pod, is a reload required?  

> Btw even the route deletion doesn't lead to a reload if DCM is enabled. All of the servers of the deleted backend will be disabled but the backend will stay in `show backend` output.

This is excellent news!  

Can you reproduce these results with non-TLS, TLS passthrough, edge-terminated, and reencrypt routes?  Since there are some nuances (`server` versus `server-template`), I'm wary of subtle differences in behavior for different route types.

Reply:
> I'm confused. I've been saying "pre-allocated" to describe the slots that we pre-allocate in haproxy.config for DCM to use to add new server endpoints to a backend. What do you mean when you say "3 pre-allocated slots"? Are these 3 pre-allocated slots for DCM, or are these 3 server lines for endpoints that are defined in haproxy.config?

Sorry, my bad. I should have called them "static" slots. The pre-allocated are dynamic slots.

> Are you saying that I can create a route with 3 endpoints (which get written to haproxy.config), and then scale in to 1 endpoint, and no reload is required?

Right, if DCM is enabled.

> If I have a route with 1 pod endpoint (written to haproxy.config), delete the pod, then recreate the pod, is a reload required?

No. The deletion of the pod will set the corresponding server to `maintenance` mode. The recreation of a pod will use the dynamic (pre-allocated) server slot to set the new IP.